src/function-builder-detectors/zisi.js

```diff
@@ -0,0 +1,65 @@
+const path = require('path')
+
+const { zipFunction, zipFunctions } = require('@netlify/zip-it-and-ship-it')
+const makeDir = require('make-dir')
+
+const { getPathInProject } = require('../lib/settings')
+const { NETLIFYDEVERR } = require('../utils/logo')
+
+const bundleFunctions = ({ config, sourceDirectory, targetDirectory, updatedPath }) => {
+  // If `updatedPath` is truthy, it means we're running the build command due
+  // to an update to a file. If that's the case, we run `zipFunction` to bundle
+  // that specific function only.
+  if (updatedPath) {
+    return zipFunction(updatedPath, targetDirectory, {
+      archiveFormat: 'none',
+      config,
+    })
+  }
+
+  return zipFunctions(sourceDirectory, targetDirectory, {
+    archiveFormat: 'none',
+    config,
+  })
+}
+
+// The function configuration keys returned by @netlify/config are not an exact
+// match to the properties that @netlify/zip-it-and-ship-it expects. We do that
+// translation here.
+const normalizeFunctionsConfig = (functionsConfig = {}) =>
+  Object.entries(functionsConfig).reduce(
+    (result, [pattern, config]) => ({
+      ...result,
+      [pattern]: {
+        externalNodeModules: config.external_node_modules,
+        ignoredNodeModules: config.ignored_node_modules,
+        nodeBundler: config.node_bundler === 'esbuild' ? 'esbuild_zisi' : config.node_bundler,
+      },
+    }),
+    {},
+  )
+
+const getTargetDirectory = async ({ errorExit }) => {
+  const targetDirectory = path.resolve(getPathInProject(['functions-serve']))
+
+  try {
+    await makeDir(targetDirectory)
+  } catch (error) {
+    errorExit(`${NETLIFYDEVERR} Could not create directory: ${targetDirectory}`)
+  }
+
+  return targetDirectory
+}
+
+module.exports = async function handler({ config, errorExit, functionsDirectory: sourceDirectory }) {
```

Comment:
At the moment we evaluate detecters based on the order of `readdirSync`:
https://github.com/netlify/cli/blob/f5b26d2295f46975194fc4c984e7695727b02029/src/utils/detect-functions-builder.js#L6

I don't think it is guaranteed to have a consistent order.

1. Should we sort the files by name?
2. Should we add a condition to not use `zip-it-and-ship-it` builder when `netlify-lambda` is used?
3. Should we add a condition to use `zip-it-and-ship-it` only when someone is using `esbuild` (I think that would minimize the risk for the PR and we can later on make this behavior the default).

Reply:
Great points.

I have done 1 and 3 in https://github.com/netlify/cli/pull/2094/commits/1b2278b755c89a57965d5f7dc2d730d366079183.

I don't think we need 2, because `netlify-lambda` will be evaluated before `zisi`, due to the alphabetical order, and if it returns a truthy value, we stop looking for the next handler (https://github.com/netlify/cli/blob/feat/serve-bundled-functions/src/utils/detect-functions-builder.js#L18).